### PR TITLE
WSE-255 Implement static host infrastructure

### PR DESCRIPTION
**Pull request by mlkirkpatrick**
This is designed to be generic and reusable with minimal effort to add new sites. As such, the list of sites and their configured parameters can all be set up from `lib/config.ts`. These values can still be overridden with namespaced context keys (e.g. `hackathon:appSourceBranch`), but having this configuration in a ts file instead of just context has several benefits. It means we get type-checking goodness so you are less likely to mistype a key while adding a new project, and the validation that appropriate props are passed down.

### Supported Sites
The initial ticket indicated that these blueprints should support "Hackathon, GIS Day, [and] Christmas card," so those are the ones currently implemented. There are other sites like Father Ted and some MARBLE sites which are built with similar infrastructure, and if we want to migrate them to use this centralized code, those will be done in future tickets.

### Syncing Site contents
The pipeline handles syncing up files from the app repository to the s3 bucket, and can even run build scripts if configured to do so (useful for SPAs.) However, pipelines can be a bit inconvenient when testing, so you can also deploy a development stack on its own. A script is included (run with `yarn sync`) that basically runs the same commands the pipeline would to sync files up to s3. This has to be done separately from deploying the stack because BucketDeployments can run into issues with file size limits.

### Transclusion Lambda
The last notable feature of this infrastructure I'd like to point out is the transclusion lambda. In essence, I discovered that the sites we wanted to support make use of server-side includes (SSI). This looks like includes inside of the HTML like this:
```<!--#include virtual="ssi/brand.bar.shtml" -->```

This Lambda@Edge will run whenever an shtml file is requested and will then recursively fetch all files which match the include pattern. It then replaces that tag with the HTML from the respective file, as fetched from the cloudfront's associated s3 bucket. The result is a flat HTML file, so we aren't forced to make any changes to how we compose these sites.  Cache times may be further optimized.

### CODE REVIEW

**Review by SteveMattison — APPROVED** [review]
Very nice.  I particularly like the config.ts and the transclusionLambda.

**Comment by mlkirkpatrick** [comment]
I added several other sites to the config so I could test out how robust the design is. (With the caveat that they might need further work since they are out-of-scope.) Found a few missing features and implemented them. One limitation that remains is that this design only supports newman smoke tests, and `library-research-award` in particular uses selenium.

Because of that, these blueprints will need some changes before this can be the one-stop-shop for static host infrastructure, but I suppose that's a problem for another time. I think it's well on its way!

**Review by jonhartzler — APPROVED** [review]

**Review by jgondron — COMMENTED** [review]
First, I do like the inclusion of config.ts. I've been wanting to explore ways to move away from using context for config, for the reasons you mention, while still allowing for overrides. This has been asked a few times in AWS' slack, and some have pointed to npm packages that do similar things, although I think the history is gone and the package name escapes me atm. In any case, glad to see someone is exploring this. 

Second, I'm willing to give this configuration a shot if others are, but I've come to the conclusion that the things that should be reused between projects are Constructs, not Stacks, and that infrastructure code (at least for service stacks) should not be co-located with other projects.

### Why constructs and not stacks:

- It's a better separation of concerns to compose Constructs into something larger (ex: marble) than it is to do so with Stacks. Otherwise you have a ton of branching logic per project to tell the stack what it should and should not include. It's similar to the typical Inheritance vs Composition argument, except in this case the inheritance is happening across multiple projects. 
- It keeps Stack counts down since you're able to compose more things into a service stack
- More things can be resolved at synth time instead of deploy time if the exposed properties are between Constructs that are composed into a Stack than if it's an exposed property of a stack (whether via exports or ssm)
- It makes it easier to use in other projects. Things that could be made into reusable Constructs in this PR that would be useful in other projects: StaticHost, StaticHostBuildAction, NewmanTestsAction, CDKDeployAction, S3SyncAction, SPARedirectLambda, etc

### Why not co-locate service stacks between projects:

These pipelines deploy infrastructure, so all of these will be watching this repo. So the main concerns are:
- This will be noisy from an approval standpoint, ex: the deluge of approvals that you see on marble any time the infra repo changes will also happen except here it will be across different project.
- It could also introduce bugs across different projects that you did not intend or fully test for, especially if you have a lot of branching logic just to support multiple projects as mentioned above.

File: README.md [hunk 1]
```diff
@@ -1,2 +1,35 @@
...
 Infrastructure for deploying static websites in a repeatable fashion to be hosted in AWS.
+
+## Deployment
+
+`env` context variable refers to the AWS account being targetted.
+`project` is an identifier for the sites that have been configured to be handled with this blueprint. Refer to [lib/config.ts](lib/config.ts) for the list of projects and their default configurations. (Update this config file if adding a new project.)
+
+### Pipeline
+```
+cdk deploy -c env=<dev|prod> -c project=<projectName> -c stackType=pipeline -c slackNotifyStackName=[stackName]
```
[the remark this answers is not in the document]
Comment (mlkirkpatrick) [reply README.md:11]: Yes, so the app synthesizes one stack instead of one for each project. From my perspective, this has advantages like the fact that synthesizing is way faster and doesn't get exponentially longer as you add on more projects. It also means you won't get error messages for irrelevant stacks, like an SSM parameter missing for a project you aren't trying to deploy anyway. (More likely to be an issue in testlibnd.)

But this hinges on the argument between reusable stacks vs components which you have pointed out. I think if they were composed as multiple stacks, I would remove the project parameter as you suggested.

**Comment by mlkirkpatrick** [comment]
@jgondron Thanks for your feedback! I can definitely see where you're coming from. I'm a bit torn on this. It's definitely more flexible to use constructs, but it is also a more complex design, and there is sort of a minimum level of variance before a complex design saves time.

### Red tape increases development time
In the case of microsites, it is a bit of a chore to create a new infrastructure repository and new stacks explicitly for that. You then are likely going to copy code from one of the other repositories, and may have to deal with version differences between the latest aws-cdk and other dependencies. (This goes back to the monorepo vs. polyrepo argument.) All the code is spread out, so it is more difficult to manage deprecations and make sure everything gets updated that needs to be.

### Reusable constructs are not necessarily easy to use
Worse yet, there are specific aspects of the stack which the pipeline assumes exist. (For instance, the parameters with the bucket name and distribution id.)
- Splitting up the pipeline and service infrastructure means you need to cross-reference the two repositories carefully when you are composing the stack. These stacks are going to look extremely similar, which is a lot of code duplication. The developer is responsible for ensuring the stack has all of the things a microsite stack needs, even though we have an established convention which could be enforced by using a one-stack structure.
- Using S3Sync as an example, it is _mostly_ generic, but makes assumptions about file types and some parameters are required. To consume this construct in their own pipeline, the developer would have to ensure their stack is exporting appropriate variables. This is easy enough to figure out by looking at the code if it's in the same project, but when you are consuming an external dependency looking through the code of the library adds weight. Ideally there should be documentation, but that also adds development effort for making any changes.
- If the constructs are to be reused, they must be retrieved somehow. If not in the same repository, that would likely be as an npm module. This adds overhead to publishing new versions, and projects that were built longer ago will drag behind and probably not be updated to the latest version of the library.

In my opinion, it is tedious to maintain separate stacks for every project when they are all highly similar as these are. Reusable constructs could help, but I don't feel as though it does _enough_ to remove the burden of adding a new project. I might be wrong; maybe it just needs to be refined and implemented just right. But having created many similar CDK apps already, I can say that I spend way more time tweaking them and trying to refine the design than I would like. When I come up with an improvement, it's too much hassle to go through the dozen other repositories which could benefit from it and make PRs, when that is technically outside the scope of the work I was given. There is something to be said about the ease of having everything in one place and modifying **one** file. It can save a lot of time, even if it does have a few drawbacks.

There's also the fact that the lambda code is going to be shared between multiple projects. I don't like copying it to each service repository (DRY principle) and keeping it in the "global blueprints" adds cross-repository dependencies if they are separate. This makes it harder to deploy.

I would be remiss if I didn't address some of the drawbacks to a shared stack, though.
- The design proposed for pipelines **does not** trigger on infrastructure changes. In practice, these changes usually require a human to identify which projects are impacted anyway. From my perspective, it is better to avoid the needless chatter and request that the developer manually release updates on the relevant pipelines when infrastructure changes are made. If they forget, it will of course be picked up when an app change is made and the pipeline runs anyway, so I don't see this as a major issue with the automation.
- Branching logic is inevitable, but one of the key components of the design is that it is _configurable_, not jury-rigged into the stack (e.g. conditions like `if (project === 'hackathon')` **must not** be used.) A stack that is carefully designed to be generic should accept a variety of inputs without breaking, and we can test different permutations. This is something you have to consider with the props in reusable constructs just the same.
- The static-host-stack attempts to provide configurability within a given pattern, but is not an appropriate choice for projects that fall outside that pattern. It doesn't address every scenario, and if the functionality you want isn't supported, you may just have to write a stack from scratch. It's the idea of solving 90% of scenarios with minimal effort as opposed to reducing effort in all scenarios but still having a higher minimum effort required.

You make some good points. My natural tendency is to think of counter-points to balance the scales and make sure we are taking everything into consideration, so hopefully this did not come across as dismissive. If this needs further discussion, this might not be the best forum for it. Regardless, it was helpful for me to write out so that I could organize my thoughts and can refer back to it later. :)

**Review by ialford — COMMENTED** [review]
I have some targeted comments, but I would also like to put in my general opinion on reusability. This feels more like a construct of set of constructs. Our original measurement for when to create a shared construct is when we noticed the same pattern of code in multiple places - and here we would see the pattern in every static host/microsite setup.

Understanding of the overhead of creating a shared set of constructs, it feels more like this would be best suited for a shared construct that could be configured to the details of a specific repo or set of repositories. I'd like to hear more from the rest of the @ndlib/cdk-core team on where they are thought wise with this.

File: lib/static-host-qa-project.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+              nodejs: '14.x',
+            },
+            commands: [
+              'npm install -g newman',
+              'echo "Ensure that the Newman spec is readable"',
+              `chmod 755 ${props.smokeTestsCollection}`,
+            ],
+          },
+          build: {
+            commands: [`newman run ${props.smokeTestsCollection} --env-var hostname=${props.hostname}`],
```
[the remark this answers is not in the document]
Comment (mlkirkpatrick) [reply lib/static-host-qa-project.ts:35]: I noted that we have examples of other collections that are written that way, but that is an easy fix by updating the collection. It's hardcoded in order to support the basic "default" collection included in the repo, whereas a custom smoke test collection just needs to follow that one rule.

But the limitation is duly noted. I considered making this more flexible if/when adding support or selenium tests.

[the remark this answers is not in the document]
Comment (ialford) [reply lib/static-host-qa-project.ts:35]: It'd also be important to note for the "next" static site, that the newman tests need to follow the specific convention.

File: lib/static-host-stack.ts [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        {
+          enabled: true,
+          expiration: cdk.Duration.days(365 * 10),
+          noncurrentVersionExpiration: cdk.Duration.days(1),
+        },
+      ],
+    })
+
+    this.bucket = new Bucket(this, 'SiteBucket', {
+      bucketName: `${this.stackName}-site-${this.account}`,
```
[the remark this answers is not in the document]
Comment (mlkirkpatrick) [reply lib/static-host-stack.ts:166]: I think when I named this I had a few considerations in mind:

- Should include stack name for correlation and to make it easy to find.
- Must be unique. Stack name should include the stage/prefix which makes it fairly unique, account id ensures that you can have the same stack in two different aws accounts and still not have conflicts.
- Granting minimum permissions to s3 (e.g. only the relevant buckets.)

So ultimately, I think I explicitly named them because I wanted to grant the build role permission _for only these buckets_. In hind-sight, it's probably sufficient to just give it permission to create as many buckets with the stack name as a prefix that it wants to. `'arn:aws:s3:::' + stackName + '*'` I can refactor this accordingly.

File: scripts/sync.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
+import AWS = require('aws-sdk')
```
[the remark this answers is not in the document]
Comment (mlkirkpatrick) [reply scripts/sync.ts:1]: Well this script is only intended to be run for local development. S3DeployAction is for use in a CodePipeline, which doesn't provide an easy way for a developer to sync up a site build to a dev stack. (Unless they happen to do all their development through the pipeline... But I prefer to deploy the service stack on its own.) Do you have any specific concerns about the script when it is used as a development tool?

Aside from that, I wasn't actually aware that S3DeployAction existed. It may prove to be a valid substitute for the pipeline-s3-sync project. I'll have to look into that some more!

File: scripts/sync.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      })
+    }
+
+    // Remove existing files first
+    execSync(`aws s3 rm s3://$DEST_BUCKET --recursive`, { stdio: 'inherit' })
+    // Copy the new build up to s3
+    execSync(`aws s3 cp --recursive "$BUILD_PATH" s3://$DEST_BUCKET --include "*" --exclude "*.shtml"`, { stdio: 'inherit' })
+    execSync(`aws s3 cp --recursive "$BUILD_PATH" s3://$DEST_BUCKET --exclude "*" --include "*.shtml" --content-type "text/html"`, { stdio: 'inherit' })
+    // Invalidate the cloudfront cache since files were updated
+    execSync(`aws cloudfront create-invalidation --distribution-id $DISTRIBUTION_ID --paths "/*"`, { stdio: 'inherit' })
```
Comment (ialford) [inline scripts/sync.ts:54]: Based on the comment that this is only for development, there shouldn't be a need to run an invalidation since the dev stacks have a cache time of `0`. 

Invalidations do eventually have a budgetary cost, and should only be run if they're actually needed (which doesn't seem to be the case for a dev stack of a static site). If you've seen where this is not the case, can you point me in that direction?

Comment (mlkirkpatrick) [reply scripts/sync.ts:54]: Good point. Will remove.

File: scripts/sync.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ialford) [reply scripts/sync.ts:1]: I don't know if I missed it somewhere, but could you add a note about the purpose of the sync script if there isn't one already?

[the remark this answers is not in the document]
Comment (mlkirkpatrick) [reply scripts/sync.ts:1]: It's mentioned in the readme. I'll add a comment to the script itself too.

[the remark this answers is not in the document]
Comment (mlkirkpatrick) [reply scripts/sync.ts:7]: FYI, I tried out S3DeployAction. It doesn't invalidate the cloudfront cache, so I believe a codebuild action is still required to do the invalidation.

It also does not let you specify patterns for files to include or custom headers. Technically that doesn't break anything here. I am setting the content-type to `text/html` for .shtml files when uploading, since s3 sets them to `application/octet-stream`. This would be problematic if not for the TransclusionLambda which triggers on shtml file requests and changes the content-type in its response.

[the remark this answers is not in the document]
Comment (ialford) [reply scripts/sync.ts:7]: Yeah, I think the Action is literally just to ensure a proper copy of the artifact to the S3 bucket. Other things probably have to be handled after the deploy (but within the same stage).

I don't think generally there's a recommendation from AWS to invalidate CloudFront on each deployment to a static site. [CloudFront Documentation](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/UpdatingExistingObjects.html) seems to suggest that Object Versioning is preferred over invalidation - although I would imagine there's some application changes that could be required to get this to work. Just another thing for us to file away in our minds.

**Comment by jgondron** [comment]
> @jgondron Thanks for your feedback! I can definitely see where you're coming from. I'm a bit torn on this. It's definitely more flexible to use constructs, but it is also a more complex design, and there is sort of a minimum level of variance before a complex design saves time.
> 
> ### Red tape increases development time
> In the case of microsites, it is a bit of a chore to create a new infrastructure repository and new stacks explicitly for that. You then are likely going to copy code from one of the other repositories, and may have to deal with version differences between the latest aws-cdk and other dependencies. (This goes back to the monorepo vs. polyrepo argument.) All the code is spread out, so it is more difficult to manage deprecations and make sure everything gets updated that needs to be.

I totally agree that red tape increases development time. This is why our general practice is to not worry about reusability on the initial implementation. It's only when you see that it can be used in another project that we should consider making a reusable component. By copying code across repos, you are implicitly acknowledging its reusability. If we would take a few hours to contribute a reusable component, we wouldn't have to copy/paste this code, making it easier for future devs. 

This isn't *really* a monorepo vs polyrepo problem. That argument will still exist independent of this. You're just attempting to solve a small part of the code reuse problem by co-locating the code, and unfortunately it ignores other parts that are still going to be copy/paste duplicates. For example, a lot of this code appears to be copied from marble-blueprints. Are we going to migrate that project's front ends into this repo? If so, what are we doing with the remaining pieces of marble? What about the parts that are still shared between this repo and the other repos such as marble, do we just continue to copy/paste those? How would we even share infra code as OS for marble if we logically group the infra code this way?

I wish this would solve a lot of the problems you describe. I'd be 100% onboard, but I don't think it will long term. Unless we somehow decide to just co-locate all of our infra code, this is really just a reshuffling of pieces related to the problem of code reuse and the problem will still exist.

> 
> ### Reusable constructs are not necessarily easy to use
> Worse yet, there are specific aspects of the stack which the pipeline assumes exist. (For instance, the parameters with the bucket name and distribution id.)
> 
> * Splitting up the pipeline and service infrastructure means you need to cross-reference the two repositories carefully when you are composing the stack. These stacks are going to look extremely similar, which is a lot of code duplication. The developer is responsible for ensuring the stack has all of the things a microsite stack needs, even though we have an established convention which could be enforced by using a one-stack structure.

> * Using S3Sync as an example, it is _mostly_ generic, but makes assumptions about file types and some parameters are required. To consume this construct in their own pipeline, the developer would have to ensure their stack is exporting appropriate variables. This is easy enough to figure out by looking at the code if it's in the same project, but when you are consuming an external dependency looking through the code of the library adds weight. Ideally there should be documentation, but that also adds development effort for making any changes.

You are identifying the potential for a static host pipeline as a reusable construct that is compatible with a reusable static host construct. I'm doing something similar in DEC right now for rails pipelines with the intent to move this into ndlib-cdk.

> * If the constructs are to be reused, they must be retrieved somehow. If not in the same repository, that would likely be as an npm module. This adds overhead to publishing new versions, and projects that were built longer ago will drag behind and probably not be updated to the latest version of the library.

True, but copy/paste is still worse, and this fundamentally assumes that this repo is solving our copy/paste problem. 

> 
> In my opinion, it is tedious to maintain separate stacks for every project when they are all highly similar as these are. Reusable constructs could help, but I don't feel as though it does _enough_ to remove the burden of adding a new project. I might be wrong; maybe it just needs to be refined and implemented just right. But having created many similar CDK apps already, I can say that I spend way more time tweaking them and trying to refine the design than I would like. When I come up with an improvement, it's too much hassle to go through the dozen other repositories which could benefit from it and make PRs, when that is technically outside the scope of the work I was given. There is something to be said about the ease of having everything in one place and modifying **one** file. It can save a lot of time, even if it does have a few drawbacks.

I think it's simply premature to say this. We are just starting out using cdk, so we've been trying a lot of different things. It will take some time to build up a reusable library where things become easy. We actually have to start doing that to get there and stop the copy/pasting.

> 
> There's also the fact that the lambda code is going to be shared between multiple projects. I don't like copying it to each service repository (DRY principle) and keeping it in the "global blueprints" adds cross-repository dependencies if they are separate. This makes it harder to deploy.

You can also provide lambda code in a module. I have not tried this yet, but I know cdk does it as part of their s3 deploy since they provide a custom resource that handles the invalidation (which is just a lambda that cloudformation will use).

> 
> I would be remiss if I didn't address some of the drawbacks to a shared stack, though.
> 
> * The design proposed for pipelines **does not** trigger on infrastructure changes. In practice, these changes usually require a human to identify which projects are impacted anyway. From my perspective, it is better to avoid the needless chatter and request that the developer manually release updates on the relevant pipelines when infrastructure changes are made. If they forget, it will of course be picked up when an app change is made and the pipeline runs anyway, so I don't see this as a major issue with the automation.

The problem with this is that you're no longer doing CI/CD. This increases the risk for one of those other projects that have not performed an exec in some time. From a site reliability standpoint, this is the greatest cause for concern for how this is organized.

> * Branching logic is inevitable, but one of the key components of the design is that it is _configurable_, not jury-rigged into the stack (e.g. conditions like `if (project === 'hackathon')` **must not** be used.) A stack that is carefully designed to be generic should accept a variety of inputs without breaking, and we can test different permutations. This is something you have to consider with the props in reusable constructs just the same.

True. The difference though is that creating a reusable construct forces you to think about this separation of concerns, and then test it. I can see here someone making a change without testing how those other stacks will be impacted. I can also see things more easily creep into this stack that violates a proper separation of concerns simply because it's easy to do so.

> * The static-host-stack attempts to provide configurability within a given pattern, but is not an appropriate choice for projects that fall outside that pattern. It doesn't address every scenario, and if the functionality you want isn't supported, you may just have to write a stack from scratch. It's the idea of solving 90% of scenarios with minimal effort as opposed to reducing effort in all scenarios but still having a higher minimum effort required.

Correct, this is why I focus more on providing constructs,  not stacks, as the building block. If someone decides that this construct would get them 90% of the way, but they need some extra pieces, they can more easily compose this construct with other things than if it's provided as a reusable stack.

> 
> You make some good points. My natural tendency is to think of counter-points to balance the scales and make sure we are taking everything into consideration, so hopefully this did not come across as dismissive. If this needs further discussion, this might not be the best forum for it. Regardless, it was helpful for me to write out so that I could organize my thoughts and can refer back to it later. :)

Not taken as dismissive. I would summarize my sentiments this way: we are still early in the process of our use and reuse of cdk code, and the feeling of things being difficult is valid. But, we've already learned a ton, and need to start applying these learned patterns in an efficient way. We just have a disagreement on how to do that. 

I should also say that my perspective is shaped by many other things that may not be transparent:

* We may want to (or be required to in the case of Marble) continue to write OS infra code that we share with other orgs. In addition to our commitments in Marble where we share with other institutions, also think about the potential for sharing with others in our org such as OIT
* You are already reusing things that we've pushed to ndlib-cdk that have made *this* project way easier that you aren't acknowledging, proving it works. We just need to spend a little extra time in creating these things, and we need to broaden our thinking on what is reusable, ex: pipelines can also be reusable
* I've been trying to resist the urge to make all of these things myself in ndlib-cdk, to allow space for others to do this. Otherwise, you get a lot of opinionated things that I've made, and then you'll have to maintain them when I transition out of this role. If there are ways that we could make contributing to ndlib-cdk easier for developers, we really should talk about that, because I do think that if we had a more complete library of things it would be much easier to stand up new projects, regardless of how the repositories are organized.

**Review by rdoughty — APPROVED** [review]

**Comment by mlkirkpatrick** [comment]
WSE met and discussed this topic. We determined as a team that we did not want to move StaticHost to ndlib-cdk at this time, largely because of the added maintenance challenges that have been referenced here. I did identify a few pieces which could be separated out for some clear wins, and added them to ndlib-cdk:

- TransclusionLambda
- SpaRedirectionLambda
- NewmanRunner (smoke tests)

There are a few other pieces that I could see moving over to constructs but have not yet, like the pipeline-s3-sync project and the context helpers. I don't have another use-case in mind for these besides MARBLE, though. If another project crops up that could benefit from that, I would suggest we add constructs to ndlib-cdk then.

For right now, though, we would like to proceed with the static-host and static-host-pipeline as is. The hope is that the simplicity lends itself to easy management for the use cases we have identified, such as microsites. If it turns out to be problematic down the line, we can always reconsider the structure and refactor as needed.

**Comment by jgondron** [comment]
A few questions, given this is the direction everyone wants to go:

1. There could be a fundamental misunderstanding on my side about what this static-host stack does, so I'll start with: Is this static-host also similar to beehive, library website, and others, and if so are we going to move those here?
2. Since this seems to be a lot of copy/pasted code from marble-blueprints, is there a plan to move its static-host into this repo? 
    a. If not, is there a plan for incorporating changes made to this repo's static-host code back into marble since they are similar patterns?
    b. If so, how do we intend to share this part of the marble infrastructure code with other institutions?
3. Is there a desire to do this for other similar patterns? Ex: We have several rails applications that we push to ECS. Do we want to move all of the infra code for ECS based services into a single repo to make the code reuse between them easier?
4. The answer to #2 can very well be no, simply because static-host is a more narrow pattern with not as much variation vs the many diff ways we might use ECS, but that then raises the question: will we have a repo for all of the dec-blueprint related infra code, except for beehive since that one will be here?
5. Are you all committing to keeping these pipelines fully executed? As someone who is trying to ensure reliability of these sites, a scenario I do not want to see:
    - We add some change `commit-A` to the static-host stack to support something we need for siteA. 
    - We click the approve button for siteA, but ignore or reject the rest since the change should not apply to the rest. 
    - This pattern goes on for many days/weeks/months, then we decide we need to push a change `commit-Z` to siteB. 
    - We push siteB forward and it breaks, although it's not immediately seen. 
    - Duty Officer/SRE/etc responds to an issue with siteB and sees that it was recently deployed with `commit-Z`. We spend time looking into why that change has caused the problem and how to fix. Except it isn't the problem. Because we haven't been continuously integrating and deploying all of the changes, `commit-A` or anything in between the two could be the actual problem.

    Note: You may see that this scenario could still exist with a reusable construct. The difference is that when we move something to ndlib-cdk, we acknowledge its potential impact on multiple downstream applications, hence the greater focus on the integrity of the code by forcing test coverage, use of semantic versioning, etc. For this scenario to happen with a managed package, someone would have to introduce the bug into ndlib-cdk (less likely given the greater testing requirements), then bump the version of ndlib-cdk in the downstream app for siteB before this problem would show. Even if the greater testing doesn't catch the bug, it's still much easier to trace the cause to the bumping of the versions and backtrack on that one project that broke. Here, there is no moving one project back temporarily since the pipelines are all pushing forward the head commit at all times. There's also less integrity in the process so it's easier for the scenario to play out. These things combined put an even greater importance on CD of all pipelines. So, I just need a commitment that we will CD all of these projects, so that we can always quickly trace the issue to the offending commit, wherever it may appear.

**Comment by mlkirkpatrick** [comment]
1. Potentially, although I don't think there is a decision around those. This is for the "simple" sites first and foremost, and if it happens to work for the rest with minimal changes then some could be consolidated.
2. I can't say. I think it could be, but its use of the foundation stack is probably reason enough to leave it as it is. The different MARBLE stacks are deeply coupled; the same cannot be said about most of our other projects (nor should they be.) Not to mention the issue of sharing the code if it was moved as you pointed out.
3. My opinion: If it is used often enough and similar enough that it makes sense. But I'm not expecting that is the case.
4. I would treat DEC like MARBLE, which is to say it's probably better to keep all the infra code together. This solution is intended for projects which can be entirely encapsulated with a static host and a pipeline; that's not the case for DEC.
5. Most problematic changes are either going to break the pipeline or break all sites. I don't really foresee any breaking changes to this stack in the future, and the config is designed so that when you are changing one project you don't impact another. The most likely culprit would be an added config parameter; if you maintain the default for all previous projects and only override it on the one you are concerned with, other sites won't break. Code review is a chance to catch any violations which would cause breaking changes.
Furthermore, I still have an expectation that any new config parameters should be tested in the unit tests. (E.g. If you add a new boolean, you must have tests for both true and false, confirming that the projects which don't use it are unimpacted by the addition.)
We also shouldn't be pushing to siteB without doing any verification to ensure it's working afterward. A breaking change for these static hosts will probably be immediately obvious as soon you pull up the site. I can see where some cache misconfiguration or something might go unnoticed, but that doesn't warrant a duty officer addressing it.
But none of that completely eliminates the risk; sometimes we skip or miss things we _should_ be doing as safeguards. It's just an explanation of why I think the risk is very low. For the remaining doubt, can you think of a scenario where there is anything in this stack we would reasonably want to change for a particular site (that isn't accounted for already with config), which would break the others?

I recognize that 2 and 4 mean we are not replacing all static hosts, so there is still some duplication. (At least 3 or 4 versions between MARBLE, DEC, usurper, and "the rest.") I can see how using a shared construct would alleviate some of this, but it's outside of scope for right now.

**Comment by ialford** [comment]
So, I've been trying to digest all of the conversation and arguments here - and having not been part of some of the other conversations that have gone on around this, I can only go with my own thoughts and the concerns raised within this thread.

TL;DR: My opinion is that this should go into `ndlib-cdk` because it fits our criteria for it.

1. It feels like there's an effort to solve only for what is being called "simple" static sites - namely that a solution for sites that are currently not in S3/CloudFront. @mlkirkpatrick you've specifically mentioned that this is not pertinent for use with Beehive, MARBLE, Usurper - and therefore I assume that things like DAVE are also out. Remix is another that will be impacted by this choice here, and fits more into the "simple" component. However, my opinion is that if they are all static sites, they should all be treated equally - branching logic in a decision tree for a new static site (do I use `static-host-blueprints` or something else) leads to more issues than no branches. So, if this doesn't fit with the items outlined, we should make it fit. Once deployed, there's little difference in the infrastructure for the MARBLE front-end against something like Code Café.

2. The term over-engineering is a misnomer, in my opinion. What is being called over-engineering (and if this is a @ndlib/wse-unit   concern, we need other voices) is, to me, holding to the standards we set when creating the `ndlib-cdk` shared library. The fact that we're bypassing the step of identifying the shared pattern by wanting to move multiple sites at once to a new infra pattern doesn't negate the fact that these are shared patterns (an S3, a CloudFront, a deploy pipeline for static assets, etc) that we've previously said should go into `ndlib-cdk` if identified. 

3. From both a reliability hat and a duty officer hat, the thought of changes coming through a pipeline at an inconsistent time is not one I relish. We've typically done one of two patterns: continuous delivery (where _every_ commit to `main` is an RC) or tag-and-release (typically through Jenkins, where we can release and re-release any specific tagged version of the codebase). This introduces a third way as I see it - infrastructure only updated when the application is updated. This means two potential sources of change in a single deploy - making traceability more difficult and raising the chance for unforeseen consequences. It's great to say that code review could catch breaking changes - and while I agree,  it's also possible for them to be missed. If this were in `ndlib-cdk` as a set of constructs, the tagged version within a specific set of blueprints would mitigate this risk by only building what we expected to build, and giving us an intentional way of rolling future changes to the constructs out in a deliberate fashion. As a duty officer, I could say that one of my first questions in a breaking change situation would be "how did this happen" followed closely by "what can we do to not let it happen again." If the answer is `ndlib-cdk`, then I believe it makes more sense to do that work now.

4. To address something specific:
>> Is there a desire to do this for other similar patterns? Ex: We have several rails applications that we push to ECS. Do we want to move all of the infra code for ECS based services into a single repo to make the code reuse between them easier?

>>> My opinion: If it is used often enough and similar enough that it makes sense. But I'm not expecting that is the case.

I'm having concerns over who is the arbiter of "when it makes sense," and why the concerns that are being outlined here on both sides wouldn't just be brought back up in that discussion, should it arise. Meanwhile, we had a very specific criteria for when we should move something into `ndlib-cdk` - code patterns across multiple services. The specifics here are definable and the subjective is not.

Some of the language used throughout this discussion has felt loaded to me, and I really want to try and get away from that. My take on the opinions has been that @jgondron and I have been asking why this doesn't belong in `ndlib-cdk`, and the answer from @mlkirkpatrick as author and the only member of WSE to participate in the thread has been "we don't see the need or what it solves."  If I am misstating this, please let me know. 

Regardless of if we went with this code as-is or migrated any/all of it to `ndlib-cdk`, there are some things I would like to see changed, such as increased documentation (think of the new developer coming in after you win the lottery) and unit testing via CI. 

>>  I can see where some cache misconfiguration or something might go unnoticed, but that doesn't warrant a duty officer addressing it.

I can tell you from experience - if there's a problem, a DO is likely to be raised on it, even if it is minor. Admittedly, our model places the DO at the frontlines of issue identification and escalation - so it might even be the case that a DO is pinged on something but it doesn't get raised with an SRE or developer until the next business day.

There's a lot in the messaging indicating that everything should work fine as long as nothing gets changed in the config. I think a very specific challenge with this approach will be: how does a reviewer know that a config change was intentional or accidental? An example might be changing an `index.shtml` to `index.html` without the actual code changing. It looks reasonable, it should work, but it would break the site. There's other theoretical situations I could put forth, but this isn't an exercise in theory-crafting.

Overall, it's not the solution as much as where it lives that is the concern in my mind. We are unnecessarily expanding our downstream impacts with this method. I also would wonder exactly how much overhead there would be in moving this to `ndlib-cdk` as the code is already written, and I imagine would require only minor revisions.

Lastly, Remix is likely to get an infrastructure and deployment overhaul in the near future. This was not accounted for in this project initially, and does not fall outside of the scope originally outlined for what this code is to be used for. However, it's another service to add into the area of concerns with future infra changes - and Remix can change a lot depending on how much Randy is updating things.

**Comment by mlkirkpatrick** [comment]
For posterity:
> My take on the opinions has been that @jgondron and I have been asking why this doesn't belong in ndlib-cdk, and the answer from @mlkirkpatrick as author and the only member of WSE to participate in the thread has been "we don't see the need or what it solves."

I firmly disagree with that. I have been trying to convey that I see the benefits, but there are also cons.

Regardless, I feel as though we need a cross-team discussion to resolve this. I don't think I should be the decision maker just because I happen to be doing the work. I'm fine with making ndlib-cdk constructs if that's what is decided. To be clear, my concerns are with the ongoing maintenance, not the initial groundwork.

**Comment by ialford** [comment]
> For posterity:
> 
> > My take on the opinions has been that @jgondron and I have been asking why this doesn't belong in ndlib-cdk, and the answer from @mlkirkpatrick as author and the only member of WSE to participate in the thread has been "we don't see the need or what it solves."
> 
> I firmly disagree with that. I have been trying to convey that I see the benefits, but there are also cons.
> 
> Regardless, I feel as though we need a cross-team discussion to resolve this. I don't think I should be the decision maker just because I happen to be doing the work. I'm fine with making ndlib-cdk constructs if that's what is decided. To be clear, my concerns are with the ongoing maintenance, not the initial groundwork.

This is a fair distinction. I should say that it's more of the value proposition overall.

**Review by ialford — APPROVED** [review]
I disagree with this as the "right" solution, but in the spirit of not letting the perfect be the enemy of the good, I am approving.

The ongoing conversations felt too emotionally taxing and did not seem to have enough participants with strong enough opinions to make continuing worthwhile.
